apps/emqx/src/emqx_router_index.erl

```diff
@@ -0,0 +1,122 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2017-2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_router_index).
+
+-behaviour(gen_server).
+
+-include("emqx.hrl").
+-include("emqx_router.hrl").
+-include("types.hrl").
+-include("logger.hrl").
+
+-define(ROUTE_INDEX_TRIE, emqx_route_index_trie).
+
+-export([start_link/0]).
+
+-export([match/1]).
+
+%% test only
+-export([peek_last_update/0]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+-spec start_link() -> startlink_ret().
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec match(emqx_types:topic()) -> [emqx_types:route()].
+match(Topic) ->
+    match_entries(Topic).
+
+-spec peek_last_update() -> maybe({write | delete, emqx_types:topic(), emqx_router:dest()}).
+peek_last_update() ->
+    gen_server:call(?MODULE, peek_last_update).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([]) ->
+    _ = emqx_trie:create_local(?ROUTE_INDEX_TRIE, [public, named_table]),
+    ok = mria:wait_for_tables([?ROUTE_TAB]),
+    % NOTE: it's unclear if we have guarantees that we won't miss any updates
+    {ok, _} = mnesia:subscribe({table, ?ROUTE_TAB, simple}),
```

Comment:
🔥 IMO great approach
For retainer indexing, we discussed with Dmitrii a possible way to "subscribe" to retainer table updates and async-ly update some FTS index. This would remove all the reindexing complexity.

But for some reason (maybe because of my bad explanatory skills) we missed this option 🤔 — just a subscription to the local table. If this approach proves to be reliable, we may improve the retainer significantly.